tools/urtrace/collector.cpp

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ * @file collector.cpp
+ *
+ * This file contains the implementation of the UR collector library. The library
+ * provides instrumentation for tracing function calls and profiling function
+ * execution time.
+ */
+
+#include <chrono>
+#include <map>
+#include <optional>
+#include <regex>
+#include <sstream>
+#include <string>
+#include <string_view>
+
+#include "logger/ur_logger.hpp"
+#include "ur_api.h"
+#include "ur_params.hpp"
+#include "xpti/xpti_trace_framework.h"
+
+constexpr uint16_t TRACE_FN_BEGIN =
+    static_cast<uint16_t>(xpti::trace_point_type_t::function_with_args_begin);
+constexpr uint16_t TRACE_FN_END =
+    static_cast<uint16_t>(xpti::trace_point_type_t::function_with_args_end);
+constexpr std::string_view UR_STREAM_NAME = "ur";
+
+logger::Logger out = logger::create_logger("collector", true);
+
+#define ARGS_ENV "UR_COLLECTOR_ARGS"
+
+enum time_unit {
+    TIME_UNIT_AUTO,
+    TIME_UNIT_NS,
+    TIME_UNIT_US,
+    TIME_UNIT_MS,
+    TIME_UNIT_S,
+    MAX_TIME_UNIT,
+};
+
+const char *time_unit_str[MAX_TIME_UNIT] = {"auto", "ns", "us", "ms", "s"};
+
+std::string time_to_str(std::chrono::nanoseconds dur, enum time_unit unit) {
+    std::ostringstream ostr;
+
+    switch (unit) {
+    case TIME_UNIT_AUTO: {
+        if (dur.count() < 1000) {
+            return time_to_str(dur, TIME_UNIT_NS);
+        }
+        if (dur.count() < 1000 * 1000) {
+            return time_to_str(dur, TIME_UNIT_US);
+        }
+        if (dur.count() < 1000 * 1000 * 1000) {
+            return time_to_str(dur, TIME_UNIT_MS);
+        }
+        return time_to_str(dur, TIME_UNIT_S);
+    } break;
+    case TIME_UNIT_NS: {
+        ostr << dur.count() << "ns";
+    } break;
+    case TIME_UNIT_US: {
+        std::chrono::duration<double, std::micro> d = dur;
+        ostr << d.count() << "us";
+    } break;
+    case TIME_UNIT_MS: {
+        std::chrono::duration<double, std::milli> d = dur;
+        ostr << d.count() << "ms";
+    } break;
+    case TIME_UNIT_S: {
+        std::chrono::duration<double, std::centi> d = dur;
```

Comment:
why `std::centi`? shouldn't it be `std::ratio<1>`?

Reply:
I was mechanically going down the list. You are right, it should be ratio<1>.

Reply:
done.